### PR TITLE
[xray] Adds a driver table.

### PR DESCRIPTION
**Pull request by elibol**
This PR adds a driver table for the new GCS, which enables cleanup functionality associated with monitoring driver death.

Some testing in `monitor_test.py` is restored, but redis sharding for xray is needed to enable remaining tests.

### CODE REVIEW

File: python/ray/monitor.py [hunk 1]
```diff
@@ -16,6 +16,8 @@
 from ray.core.generated.DriverTableMessage import DriverTableMessage
 from ray.core.generated.GcsTableEntry import GcsTableEntry
 from ray.core.generated.HeartbeatTableData import HeartbeatTableData
+from ray.core.generated.DriverTableData import DriverTableData
+from ray.core.generated.ray.protocol.Task import Task
```
Comment (robertnishihara) [inline python/ray/monitor.py]: this should be in `ray.gcs_utils`

File: python/ray/monitor.py [hunk 2]
```diff
@@ -42,12 +44,17 @@
...
+XRAY_DRIVER_CHANNEL = b"7"
+
 # common/redis_module/ray_redis_module.cc
 OBJECT_INFO_PREFIX = b"OI:"
 OBJECT_LOCATION_PREFIX = b"OL:"
 TASK_TABLE_PREFIX = b"TT:"
 DB_CLIENT_PREFIX = b"CL:"
 DB_CLIENT_TABLE_NAME = b"db_clients"
+XRAY_TASK_TABLE_PREFIX = b"TASK:"
+XRAY_OBJECT_TABLE_PREFIX = b"OBJECT:"
```
Comment (robertnishihara) [inline python/ray/monitor.py]: these things are already in `ray.gcs_utils`

File: python/ray/monitor.py [hunk 3]
```diff
@@ -468,6 +475,77 @@ def driver_removed_handler(self, unused_channel, data):
...
+        """Remove this driver's object/task entries from redis.
+
+        Removes control-state entries of all tasks and task return
+        objects belonging to the driver.
+        """
+        redis = self.state.redis_client
+
+        task_table_infos = {}  # task id -> TaskInfo
+        for key in redis.scan_iter(match=XRAY_TASK_TABLE_PREFIX + b"*"):
+            entry = redis.get(key)
```
Comment (robertnishihara) [inline python/ray/monitor.py]: @concretevitamin, this will interact with automatic background flushing, right?

We should include a TODO here to address that (e.g., what if `key` gets flushed in the background while this is happening).

File: python/ray/monitor.py [hunk 4]
```diff
@@ -468,6 +475,77 @@ def driver_removed_handler(self, unused_channel, data):
...
+            task = Task.GetRootAsTask(entry, 0)
```
Comment (robertnishihara) [inline python/ray/monitor.py]: Can you actually parse it this way? Don't you need to parse it as a GcsTableEntry first?

File: python/ray/monitor.py [hunk 5]
```diff
@@ -468,6 +475,77 @@ def driver_removed_handler(self, unused_channel, data):
...
+        driver_task_ids = task_table_infos.keys()
+
+        # Get the list of objects returned by driver tasks.
+        driver_object_ids = []
+        for task_info in task_table_infos.values():
+            driver_object_ids.extend([
+                task_info.Returns(i) for i in range(task_info.ReturnsLength())
+            ])
+
+        # TODO: Remove only objects in GCS that originated from this driver.
```
Comment (robertnishihara) [inline python/ray/monitor.py]: Are you saying we don't have enough information in the GCS to flush objects created by ray.put? From the object ID, we can determine the task ID, and from that we can determine the Driver ID, right? So I think we should be able to flush them already.

File: python/ray/monitor.py [hunk 6]
```diff
@@ -468,6 +475,77 @@ def driver_removed_handler(self, unused_channel, data):
...
+                format(num_deleted))
+        if num_deleted != len(keys):
+            log.warning(
+                "Failed to remove {} relevant "
+                "redis entries.".format(len(keys) - num_deleted))
+
+    def xray_driver_removed_handler(self, unused_channel, data):
+        """Handle a notification that a driver has been removed.
+        """
+        # TODO: Migrate this to raylet monitor.
```
Comment (robertnishihara) [inline python/ray/monitor.py]: Put your name in the TODO so that people know who wrote it.

However, do we actually want this in the raylet monitor? The raylet monitor is intended for detecting liveness of machines, whereas this functionality is more about resource management and cleanup.

File: python/ray/monitor.py [hunk 7]
```diff
@@ -478,6 +481,83 @@ def driver_removed_handler(self, unused_channel, data):
...
+        log.info(
+            "Removed {} dead redis entries of the driver from redis.".
+                format(num_deleted))
+        if num_deleted != len(keys):
+            log.warning(
+                "Failed to remove {} relevant "
+                "redis entries.".format(len(keys) - num_deleted))
+
+    def xray_driver_removed_handler(self, unused_channel, data):
+        """Handle a notification that a driver has been removed.
```
Comment (robertnishihara) [inline python/ray/monitor.py]: document args

File: src/common/redis_module/ray_redis_module.cc [hunk 8]
```diff
@@ -120,6 +120,8 @@ RedisModuleKey *OpenPrefixedKey(RedisModuleCtx *ctx,
       << "This table has no prefix registered";
   RAY_CHECK(prefix >= TablePrefix::MIN && prefix <= TablePrefix::MAX)
       << "Prefix must be a valid TablePrefix";
+  // Note(hme): The use of EnumNameTablePrefix directly breaks the redis key
+  // naming convention.
```
Comment (robertnishihara) [inline src/common/redis_module/ray_redis_module.cc]: Can you explain this comment?

File: src/ray/gcs/format/gcs.fbs [hunk 9]
```diff
@@ -160,3 +163,10 @@ table HeartbeatTableData {
   resources_total_label: [string];
   resources_total_capacity: [double];
 }
+
+table DriverTableData {
+  // The driver ID.
+  driver_id: string;
+  // Whether it's dead.
+  is_dead: bool;
+}
```
Comment (robertnishihara) [inline src/ray/gcs/format/gcs.fbs]: newline

File: src/ray/gcs/tables.h [hunk 10]
```diff
@@ -289,6 +289,27 @@ class HeartbeatTable : public Table<ClientID, HeartbeatTableData> {
...
+class DriverTable : public Log<JobID, DriverTableData> {
+ public:
+  DriverTable(const std::shared_ptr<RedisContext> &context, AsyncGcsClient *client)
+      : Log(context, client) {
+    pubsub_channel_ = TablePubsub::DRIVER;
+    prefix_ = TablePrefix::DRIVER;
+  };
+  virtual ~DriverTable() {}
+
+  Status AppendDriverData(UniqueID driver_id, bool is_dead) {
```
Comment (robertnishihara) [inline src/ray/gcs/tables.h]: Should be JobID instead of UniqueID

File: src/ray/raylet/node_manager.cc [hunk 11]
```diff
@@ -153,12 +153,33 @@ ray::Status NodeManager::RegisterGcs() {
...
       }));
 
+  // Subscribe to driver table updates.
+  const auto driver_table_handler = [this](
+      gcs::AsyncGcsClient *client, const ClientID &client_id,
+      const std::vector<DriverTableDataT> &driver_data) {
+    HandleDriverTableUpdate(client_id, driver_data);
+  };
+  RAY_RETURN_NOT_OK(gcs_client_->driver_table().Subscribe(
+      UniqueID::nil(), UniqueID::nil(), driver_table_handler, nullptr));
```
Comment (robertnishihara) [inline src/ray/raylet/node_manager.cc]: The first one should be `JobID::nil()`

File: test/monitor_test.py [hunk 12]
```diff
@@ -44,6 +44,8 @@ def Driver(success):
             # Two new objects.
             ray.get(ray.put(1111))
             ray.get(ray.put(1111))
+            # Test is flaky without calls to sleep.
```
Comment (robertnishihara) [inline test/monitor_test.py]: What are the sleeps needed for? Can we get rid of them?

File: test/monitor_test.py [hunk 13]
```diff
@@ -79,23 +83,43 @@ def f():
...
         ray.worker.cleanup()
         subprocess.Popen(["ray", "stop"]).wait()
 
+    @unittest.skipIf(
+        os.environ.get('RAY_USE_NEW_GCS', False),
+        "Failing with the new GCS API.")
+    def testTaskCleanupOnDriverExitSingleRedisShard(self):
+        self._testCleanupOnDriverExit(num_redis_shards=1, test_object_cleanup=False)
+
+    @unittest.skipIf(
```
Comment (robertnishihara) [inline test/monitor_test.py]: I think we should be able to re-enable this.

File: python/ray/monitor.py (hunk 5 above)
Comment (robertnishihara) [reply python/ray/monitor.py]: Also, let's make sure that we're testing cleanup for `ray.put` objects.

File: test/monitor_test.py [hunk 14]
```diff
@@ -79,23 +83,43 @@ def f():
         ray.init(redis_address=redis_address)
         # The assertion below can fail if the monitor is too slow to clean up
         # the global state.
-        self.assertEqual((0, 1), StateSummary()[:2])
+
+        if test_object_cleanup:
```
Comment (robertnishihara) [inline test/monitor_test.py]: Can you explain this if statement?

File: test/monitor_test.py (hunk 12 above)
Comment (elibol) [reply test/monitor_test.py]: The test sometimes fails without the wait in place.

File: python/ray/monitor.py (hunk 5 above)
Comment (elibol) [reply python/ray/monitor.py]: The object ID does not contain the task ID.

File: src/common/redis_module/ray_redis_module.cc (hunk 8 above)
Comment (elibol) [reply src/common/redis_module/ray_redis_module.cc]: Current redis keys have the naming convention `TASK:<TASK_ID>`. With this change in place, keys are named `TASK<TASK_ID>`.

File: test/monitor_test.py (hunk 13 above)
Comment (elibol) [reply test/monitor_test.py]: I believe re-enabling this entails a substantial amount of work. It might make sense to re-enable as a separate PR.

File: python/ray/monitor.py (hunk 5 above)
Comment (robertnishihara) [reply python/ray/monitor.py]: What I mean is that the object ID itself tells you the task ID, right? https://github.com/ray-project/ray/blob/aa423318442c18e3e73b6125a5c5237258d60ff0/src/ray/id.cc#L166-L174

File: test/monitor_test.py (hunk 12 above)
Comment (robertnishihara) [reply test/monitor_test.py]: Why does it fail, does this mean the cleanup implementation is buggy?

Comment (elibol) [reply test/monitor_test.py]: `ray.get` completes before information about the object's location is propagated to the GCS, so the count returned by `StateSummary()` immediately after a `ray.get` is incorrect.

To compare: If the object was remote, then `ray.get` would only complete after data propagates to the GCS: The associated `pull` for `ray.get` depends on information in the GCS.

File: python/ray/monitor.py (hunk 5 above)
Comment (elibol) [reply python/ray/monitor.py]: Excellent point. I was not aware of this method.

File: python/ray/monitor.py (hunk 3 above)
Comment (concretevitamin) [reply python/ray/monitor.py]: Right.  I think this just needs a simple catch to test whether `redis.get(key)` returns None.

File: python/ray/monitor.py [hunk 15]
```diff
@@ -36,6 +36,9 @@
 # xray heartbeats
 XRAY_HEARTBEAT_CHANNEL = b"6"
 
+# xray driver updates
+XRAY_DRIVER_CHANNEL = b"8"
```
Comment (robertnishihara) [inline python/ray/monitor.py]: this should go in `gcs_utils.py` (same with `XRAY_HEARTBEAT_CHANNEL`)

Actually, it should be `ray.gcs_utils.TablePubsub.DRIVER`

File: python/ray/monitor.py [hunk 16]
```diff
@@ -478,6 +481,103 @@ def driver_removed_handler(self, unused_channel, data):
...
+        Removes control-state entries of all tasks and task return
+        objects belonging to the driver.
+
+        Args:
+            driver_id: The driver id.
+        """
+        redis = self.state.redis_clients[0]
+
+        XRAY_TASK_TABLE_PREFIX = ray.gcs_utils.TablePrefix_RAYLET_TASK_string\
+            .encode("utf-8")
```
Comment (robertnishihara) [inline python/ray/monitor.py]: `.encode("ascii")`, same below

File: python/ray/monitor.py [hunk 17]
```diff
@@ -478,6 +481,103 @@ def driver_removed_handler(self, unused_channel, data):
...
+
+        Removes control-state entries of all tasks and task return
+        objects belonging to the driver.
+
+        Args:
+            driver_id: The driver id.
+        """
+        redis = self.state.redis_clients[0]
+
+        XRAY_TASK_TABLE_PREFIX = ray.gcs_utils.TablePrefix_RAYLET_TASK_string\
```
Comment (robertnishihara) [inline python/ray/monitor.py]: We should really enforce this in our linting, but let's avoid using `\` and instead write this like 

```python
XRAY_TASK_TABLE_PREFIX = ray.gcs_utils.TablePrefix_RAYLET_TASK_string.encode(
    "ascii")

# or 

XRAY_TASK_TABLE_PREFIX = (
    ray.gcs_utils.TablePrefix_RAYLET_TASK_string.encode("ascii"))
```

Comment (robertnishihara) [reply python/ray/monitor.py]: Let's avoid using all caps for temporary variables

File: src/common/redis_module/ray_redis_module.cc (hunk 8 above)
Comment (robertnishihara) [reply src/common/redis_module/ray_redis_module.cc]: Anyway, is the naming convention important? Or can we just consider the naming convention to be changed to `TASK<TASK_ID>`? If so, we can remove this comment.

File: python/ray/monitor.py (hunk 16 above)
Comment (elibol) [reply python/ray/monitor.py]: Why `ascii` and not `utf-8`?

File: src/common/redis_module/ray_redis_module.cc (hunk 8 above)
Comment (elibol) [reply src/common/redis_module/ray_redis_module.cc]: I'll remove the comment. One drawback is that we no longer have an abstract way to split name from id, like:

```
for key in redis.scan_iter("*"):
  item_name, item_id = key.split(":")
```

File: python/ray/monitor.py [hunk 18]
```diff
@@ -34,7 +34,11 @@
 DRIVER_DEATH_CHANNEL = b"driver_deaths"
 
 # xray heartbeats
-XRAY_HEARTBEAT_CHANNEL = b"6"
+XRAY_HEARTBEAT_CHANNEL = str(
+    ray.gcs_utils.TablePubsub.HEARTBEAT).encode("ascii")
```
Comment (robertnishihara) [inline python/ray/monitor.py]: Why are you applying `str` here? That will give something like `"b'6'"` which is almost certainly not what we want.

File: src/ray/gcs/format/gcs.fbs [hunk 19]
```diff
@@ -49,6 +52,11 @@ table ObjectTableData {
   is_eviction: bool;
   // The number of times this object has been evicted from this node so far.
   num_evictions: int;
+  // Whether this object was created by ray.put.
+  is_put: bool;
```
Comment (robertnishihara) [inline src/ray/gcs/format/gcs.fbs]: What do we need this for?

File: src/ray/gcs/format/gcs.fbs [hunk 20]
```diff
@@ -49,6 +52,11 @@ table ObjectTableData {
...
+  // The task that created this object.
+  task_id: string;
```
Comment (robertnishihara) [inline src/ray/gcs/format/gcs.fbs]: This shouldn't be necessary because the task ID is encoded in the object ID.

File: python/ray/monitor.py (hunk 18 above)
Comment (elibol) [reply python/ray/monitor.py]: Can you explain?

File: src/ray/gcs/format/gcs.fbs (hunk 19 above)
Comment (elibol) [reply src/ray/gcs/format/gcs.fbs]: Based on discussions with @concretevitamin's, this mitigates the need to deduplicate object ids in `monitor.py`.

File: src/ray/gcs/format/gcs.fbs (hunk 20 above)
Comment (elibol) [reply src/ray/gcs/format/gcs.fbs]: Indeed, the task id is computed from the object id in `ObjectDirectory`. How do you propose we compute the task id in `monitor.py`?

File: python/ray/monitor.py (hunk 18 above)
Comment (robertnishihara) [reply python/ray/monitor.py]: Oh, I misread the parentheses actually. I think what you have is correct.

File: src/ray/gcs/format/gcs.fbs (hunk 20 above)
Comment (robertnishihara) [reply src/ray/gcs/format/gcs.fbs]: I think we want to expose this function to Python https://github.com/ray-project/ray/blob/a8517cc82a35e72f085884f8c9e3f9c7fb63f178/src/ray/id.h#L81

File: python/ray/monitor.py [hunk 21]
```diff
@@ -478,6 +482,104 @@ def driver_removed_handler(self, unused_channel, data):
...
+        redis = self.state.redis_clients[0]
+
+        xray_task_table_prefix = (
+            ray.gcs_utils.TablePrefix_RAYLET_TASK_string.encode("ascii"))
+        xray_object_table_prefix = (
+            ray.gcs_utils.TablePrefix_OBJECT_string.encode("ascii"))
+
+        task_table_infos = {}  # task id -> TaskInfo
+        for key in redis.scan_iter(match=xray_task_table_prefix + b"*"):
+            entry = redis.get(key)
```
Comment (robertnishihara) [inline python/ray/monitor.py]: Instead of directly using `redis.get` we should probably use the `RAY.TABLE_LOOKUP` interface for doing this. We should also use `state._execute_command` to execute Redis commands so that they go to the correct shard.

File: src/ray/gcs/format/gcs.fbs (hunk 20 above)
Comment (elibol) [reply src/ray/gcs/format/gcs.fbs]: Should this be exposed via `local_scheduler_extension.cc`?

File: python/ray/monitor.py (hunk 21 above)
Comment (robertnishihara) [reply python/ray/monitor.py]: Similarly `state._keys` instead of `redis.scan_iter`.

Though we should probably replace the `keys` in https://github.com/ray-project/ray/blob/8e687cbc9838fce0b1d8190c7851b7d573562487/python/ray/experimental/state.py#L171 with `scan_iter` like you're using here.

File: python/ray/monitor.py [hunk 22]
```diff
@@ -478,6 +482,104 @@ def driver_removed_handler(self, unused_channel, data):
...
+                task_info.Returns(i) for i in range(task_info.ReturnsLength())
+            ])
+
+        # Also record all the ray.put()'d objects.
+        all_put_objects = []
+        for key in redis.scan_iter(match=xray_object_table_prefix + b"*"):
+            entry = redis.zrange(key, 0, 0)
+            if entry is None:
+                continue
+            assert len(entry) == 1
```
Comment (robertnishihara) [inline python/ray/monitor.py]: I suspect this `assert` will be False a lot of the time (e.g., for an object that was transferred to a different machine or was evicted).

Comment (robertnishihara) [reply python/ray/monitor.py]: Oh, I see you're using `zrange(key, 0, 0)`. Maybe instead we should use `RAY.TABLE_LOOKUP`.

File: python/ray/monitor.py (hunk 21 above)
Comment (elibol) [reply python/ray/monitor.py]: This might make sense to do in a subsequent PR, when we enable the shard-related tests in `monitor_test.py`.

File: python/ray/monitor.py (hunk 22 above)
Comment (elibol) [reply python/ray/monitor.py]: I think this type of refactoring makes sense to apply to the entire module in a subsequent PR, so that the module implementation remains consistent.

File: src/ray/gcs/format/gcs.fbs (hunk 19 above)
Comment (robertnishihara) [reply src/ray/gcs/format/gcs.fbs]: Ok, I see how it's being used now, but I don't think it is worth adding the additional information to the GCS. Since we're already pulling down all of the task table, we can just look up the relevant task and check the driver to see if the object corresponds to the given job.

File: python/ray/monitor.py [hunk 23]
```diff
@@ -478,6 +482,104 @@ def driver_removed_handler(self, unused_channel, data):
 
         self._clean_up_entries_for_driver(driver_id)
 
+    def _xray_clean_up_entries_for_driver(self, driver_id):
```
Comment (robertnishihara) [inline python/ray/monitor.py]: I wonder if these methods should just call `state.task_table()` and `state.object_table()`? What would be the downside of doing that?

File: src/ray/gcs/format/gcs.fbs (hunk 19 above)
Comment (elibol) [reply src/ray/gcs/format/gcs.fbs]: I agree that we can remove this from the table. I'm going to deduplicate the object ids instead of checking for `is_put`.

File: python/ray/monitor.py (hunk 21 above)
Comment (robertnishihara) [reply python/ray/monitor.py]: Hm, well it's a simple change (just a few substitutions), and then it will automatically continue to work when sharding is implemented.

File: python/ray/monitor.py (hunk 23 above)
Comment (elibol) [reply python/ray/monitor.py]: It looks like it would be more costly, but it might make sense as a separate PR for long-term consistency.

File: python/ray/monitor.py (hunk 22 above)
Comment (elibol) [reply python/ray/monitor.py]: I replaced this loop with a call to `self.state.object_table()`.

File: python/ray/monitor.py (hunk 23 above)
Comment (elibol) [reply python/ray/monitor.py]: Done.

File: src/ray/raylet/node_manager.cc [hunk 24]
```diff
@@ -460,6 +482,10 @@ void NodeManager::ProcessClientMessage(
 
       // Since some resources may have been released, we can try to dispatch more tasks.
       DispatchTasks();
+    } else {
```
Comment (robertnishihara) [inline src/ray/raylet/node_manager.cc]: There were a bunch of changes to this part of the code in #2362. I think now the right way to get the driver ID is shown in 

https://github.com/ray-project/ray/blob/ccb9a27393022b3e44fa7e073128c752ef119664/src/ray/raylet/node_manager.cc#L522-L526

File: src/ray/object_manager/object_manager.cc [hunk 25]
```diff
@@ -90,6 +94,10 @@ void ObjectManager::NotifyDirectoryObjectAdd(const ObjectInfoT &object_info) {
   local_objects_[object_id] = object_info;
   ray::Status status =
       object_directory_->ReportObjectAdded(object_id, client_id_, object_info);
+}
+
+void ObjectManager::HandleUnfulfilledPushRequests(const ObjectInfoT &object_info) {
+  ObjectID object_id = ObjectID::from_binary(object_info.object_id);
```
Comment (robertnishihara) [inline src/ray/object_manager/object_manager.cc]: Minor, but I think this can be `const ObjectID`. This comment applies in a ton of places (we haven't been very religious about it, but I think it is preferable to do that).

File: src/ray/object_manager/object_manager.h [hunk 26]
```diff
@@ -267,6 +267,10 @@ class ObjectManager : public ObjectManagerInterface {
   /// Register object remove with directory.
   void NotifyDirectoryObjectDeleted(const ObjectID &object_id);
 
+  /// Handle any push requests that were made before an object was available.
+  /// This is invoked when an "object added" notification is received from the store.
+  void HandleUnfulfilledPushRequests(const ObjectInfoT &object_info);
```
Comment (robertnishihara) [inline src/ray/object_manager/object_manager.h]: This seems unrelated to the driver table and to cleaning up the GCS, is that right?

Same with all of the object manager changes.

File: src/common/lib/python/common_extension.cc [hunk 27]
```diff
@@ -907,3 +907,12 @@ PyObject *check_simple_value(PyObject *self, PyObject *args) {
   }
   Py_RETURN_FALSE;
 }
+
+PyObject *compute_task_id(PyObject *self, PyObject *args) {
+  ObjectID object_id;
+  if (!PyArg_ParseTuple(args, "O&", &PyObjectToUniqueID, &object_id)) {
+    return NULL;
+  }
+  TaskID task_id = ray::ComputeTaskId(object_id);
```
Comment (robertnishihara) [inline src/common/lib/python/common_extension.cc]: can be `const TaskID`

File: src/ray/gcs/client.cc [hunk 28]
```diff
@@ -15,7 +15,10 @@ AsyncGcsClient::AsyncGcsClient(const ClientID &client_id, CommandType command_ty
   task_table_.reset(new TaskTable(context_, this, command_type));
   raylet_task_table_.reset(new raylet::TaskTable(context_, this, command_type));
   task_reconstruction_log_.reset(new TaskReconstructionLog(context_, this));
-  heartbeat_table_.reset(new HeartbeatTable(context_, this));
+  // TODO(hme): Eventually move HeartbeatTable to _context to avoid
+  // overloading the primary shard.
+  heartbeat_table_.reset(new HeartbeatTable(primary_context_, this));
```
Comment (robertnishihara) [inline src/ray/gcs/client.cc]: We probably want to revert this change to the heartbeat_table, right?

File: src/ray/gcs/client.h [hunk 29]
```diff
@@ -90,6 +90,10 @@ class RAY_EXPORT AsyncGcsClient {
   std::unique_ptr<RedisAsioClient> asio_subscribe_client_;
   // The following context writes everything to the primary shard
   std::shared_ptr<RedisContext> primary_context_;
+  // TODO(hme): Move HeartbeatTable to data shard to avoid
+  // overloading the primary shard.
+  std::unique_ptr<HeartbeatTable> heartbeat_table_;
```
Comment (robertnishihara) [inline src/ray/gcs/client.h]: revert changes to the Heartbeat table? Or can you explain these changes?

File: python/ray/monitor.py [hunk 30]
```diff
@@ -488,6 +492,106 @@ def driver_removed_handler(self, unused_channel, data):
...
+        xray_task_table_prefix = (
+            ray.gcs_utils.TablePrefix_RAYLET_TASK_string.encode("ascii"))
+        xray_object_table_prefix = (
+            ray.gcs_utils.TablePrefix_OBJECT_string.encode("ascii"))
+
+        task_table_infos = {}  # task id -> TaskInfo
+        task_table_objects = self.state.task_table()
+        driver_id_hex = binary_to_hex(driver_id)
+        for task_id_hex in task_table_objects:
+            if len(task_table_objects[task_id_hex]) == 0:
```
Comment (robertnishihara) [inline python/ray/monitor.py]: instead of indexing into `task_table_objects` here, let's change the for loop to be `for task_id_hex, task_info in self.state.task_table().items()` or something like that

File: python/ray/monitor.py [hunk 31]
```diff
@@ -488,6 +492,106 @@ def driver_removed_handler(self, unused_channel, data):
...
+                continue
```
Comment (robertnishihara) [inline python/ray/monitor.py]: It's probably ok to assert that this does not happen instead of continuing

File: python/ray/monitor.py [hunk 32]
```diff
@@ -488,6 +492,106 @@ def driver_removed_handler(self, unused_channel, data):
...
+            task_table_object = task_table_objects[task_id_hex][0]['TaskSpec']
+            task_driver_id_hex = task_table_object['DriverID']
+            if driver_id_hex != task_driver_id_hex:
```
Comment (robertnishihara) [inline python/ray/monitor.py]: combine this line and above into a single line

File: python/ray/monitor.py [hunk 33]
```diff
@@ -488,6 +492,106 @@ def driver_removed_handler(self, unused_channel, data):
...
+            task_table_infos[task_id_hex] = task_table_object
+        driver_task_id_bins = [
+            hex_to_binary(task_id_hex)
+            for task_id_hex in task_table_infos.keys()
+        ]
+
+        # Get the list of objects returned by driver tasks.
+        driver_object_ids = set()
+        for task_info in task_table_infos.values():
+            driver_object_ids |= {
```
Comment (robertnishihara) [inline python/ray/monitor.py]: what is `|=`?

File: python/ray/monitor.py [hunk 34]
```diff
@@ -488,6 +492,106 @@ def driver_removed_handler(self, unused_channel, data):
...
+                object_id.id()
+                for object_id in task_info['ReturnObjectIDs']
+            }
+
+        # Also record all the ray.put()'d objects.
+        all_put_objects = []
+        object_table_objects = self.state.object_table()
+        for object_id in object_table_objects:
```
Comment (robertnishihara) [inline python/ray/monitor.py]: loop over keys and values to avoid unnecessary indexing

File: python/ray/monitor.py [hunk 35]
```diff
@@ -488,6 +492,106 @@ def driver_removed_handler(self, unused_channel, data):
...
+            if len(object_table_objects[object_id]) == 0:
+                continue
```
Comment (robertnishihara) [inline python/ray/monitor.py]: can probably assert that len > 0

File: python/ray/monitor.py [hunk 36]
```diff
@@ -488,6 +492,106 @@ def driver_removed_handler(self, unused_channel, data):
...
+            object_id_bin = object_id.id()
+            task_id_bin = ray.local_scheduler.compute_task_id(object_id).id()
+            all_put_objects.append((object_id_bin, task_id_bin))
```
Comment (robertnishihara) [inline python/ray/monitor.py]: this seems to be all objects, not just all put objects, right?

File: python/ray/monitor.py [hunk 37]
```diff
@@ -488,6 +492,106 @@ def driver_removed_handler(self, unused_channel, data):
...
+
+        # Keep objects from relevant tasks.
+        driver_task_ids_set = set(driver_task_id_bins)
+        for object_id, task_id in all_put_objects:
```
Comment (robertnishihara) [inline python/ray/monitor.py]: If this actually covers all objects and not just put objects, then maybe we don't need to loop through the task return values above, is that right?

**Review by robertnishihara — COMMENTED** [review]
Thanks @elibol, this looks pretty good. I left some questions/comments. @eric-jj want to take a look also?

**Comment by elibol** [comment]
jenkins, retest this please

**Comment by elibol** [comment]
jenkins, retest this please

File: src/ray/object_manager/object_manager.h (hunk 26 above)
Comment (elibol) [reply src/ray/object_manager/object_manager.h]: Yes, it's unrelated. This just improves the readability of some (at the time) recent changes to the object manager. We could make this a separate PR, but it's a relatively benign change.

**Review by robertnishihara — APPROVED** [review]

**Comment by robertnishihara** [comment]
Looks good to me, I'll merge this once the tests pass.

**Comment by robertnishihara** [comment]
Nice work @elibol!

**Comment by zhijunfu** [comment]
@elibol This is nice. I'm working on job management on ray, and @robertnishihara pointed that I could leverage this for [job cancellation](https://github.com/ray-project/ray/issues/2748#issuecomment-416470614). 

I saw the code for doing actually cleanup at driver death is currently marked in [TODO](https://github.com/ray-project/ray/blob/90ae8f11df178f56f39980e7a97eb96c5e4c2124/src/ray/raylet/node_manager.cc#L214-L224)  Wondering if you have plans to finish this part sometime soon?  Or if you're busy with other stuff, maybe I could take some time to work on this?

**Comment by elibol** [comment]
@zhijunfu that sounds great. Please do have a look into it.

**Comment by robertnishihara** [comment]
Thanks @zhijunfu, could you share an outline of what you plan on implementing?

**Comment by zhijunfu** [comment]
@elibol Yes will look into it. thanks.

**Comment by robertnishihara** [comment]
@zhijunfu the driver cleanup should be implemented separately (and before) implementing the job submission, right?

**Comment by zhijunfu** [comment]
@robertnishihara yes, the cleanup code is mostly done. Will do some more testing and then submit it to internal review, will send a PR after that:)
